### ISSUE TITLE
VSTHRD002: Don't warn when getting Result on completed task

### ISSUE THREAD

**Issue by KristianWedberg** [microsoft/vs-threading#301]
#### Bug description
VSTHR002 warns even though the task has been checked for completion or have been awaited.

It should _not_ warn in these cases, since the task is provably completed, and `Result` therefore
is safe to use. 

Not warning is important since the below is a common performance optimization - retrieving
`Result` on a completed task is roughly 2x faster than awaiting the completed task.

**Edited:** As @AArnott points out: guarding with `IsCompleted` will wrap completed exceptions
in an `AggregateException`. To avoid this, instead guard with `IsCompletedSuccessfully` (`ValueTask` 
or .Net Core 2.x) or `task.Status == TaskStatus.RanToCompletion` (.Net Framework), since 
a successfully completed task won't have any exceptions.

#### Repro steps
```C#
var task = MethodAsync();
if (!task.IsCompleted)
    await task.ConfigureAwait(false);
var useTaskResult = task.Result;
```

#### Expected behavior
VSTHR002 does _not_ trigger when the task has been checked as completed via:
- `IsCompleted`
- `IsCanceled`
- `IsFaulted`
- `IsCompletedSuccessfully` (`ValueTask` or .Net Core 2.x)
- awaited

#### Actual behavior
VSTHR002 warns even though the task has been checked for completion or have been awaited.

- Version used: 15.7
- Application (if applicable): .Net Framework 4.6.1, C#7.2

**Comment by AArnott** [comment]
Yes, I agree we should be smarter for this analyzer and not report where it's clear from code inspection of the method that the task has already completed. We've done *some* work in this area already (e.g. #219). What you ask for generally requires execution flow analysis, which the Roslyn API does not yet offer but I hear is coming. Our plan has been to address it then. However some common cases like what you pointed out could probably be hard-coded to be recognized and suppress the warning.

> Not warning is important since the below is a common performance optimization - retrieving
Result on a completed task is roughly 2x faster than awaiting the completed task.

I think you should double-check your perf test here, because the C# code you wrote is virtually equivalent to the code that the C# compiler itself emits for an await. It literally checks `IsCompleted` first and accesses the `Result` synchronously when it is already completed and only schedules the continuation when it is not complete. 
Also note that your "optimization" does change behavior in the failure case, leading to non-deterministic results. If `task.IsCompleted` is true when you check it, you'll call `task.Result` and throw an `AggregateException`. But when `task.IsCompleted` is false when you check it, then when you await you'll end up throwing the inner exception of the `AggregateException` (most likely a different exception type). Your race condition could lead to your exception handlers not firing consistently.
While you *could* solve the race condition by changing your `task.Result` call to `task.GetAwaiter().GetResult()` (which only differs in which exception is thrown), I suggest you use *await* consistently and trust the compiler to do that if check for you.

**Comment by KristianWedberg** [comment]
@AArnott All good points, my repo above was minimalistic.

On performance, the differences certainly only matter with high volume and when a completed task is much (in my case +10,000 times) more common than an incomplete task. In my tests below, without `ConfigureAwait()` the difference is a negligible **14%**. In my library though I use `ConfigureAwait(false)` everywhere, which makes it about **400%** slower, and the ability to avoid that is (for me) hard to ignore. Yes one can add additional APIs to partly mitigate it, but multiple APIs for the same thing has its own drawbacks. If I'm missing something obvious though, I would be all ears :-)

``` ini
BenchmarkDotNet=v0.10.14, OS=Windows 7 SP1 (6.1.7601.0)
Intel Core i7-4770K CPU 3.50GHz (Haswell), 1 CPU, 4 logical and 4 physical cores
Frequency=3417031 Hz, Resolution=292.6517 ns, Timer=TSC
  [Host]     : .NET Framework 4.7.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2650.0
  DefaultJob : .NET Framework 4.7.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2650.0
```
|                       Method |      Mean |     Error |    StdDev | Scaled |
|----------------------------- |----------:|----------:|----------:|-------:|
| AwaitTaskConfigureAwaitAsync | 10.766 ns | 0.0202 ns | 0.0189 ns |   4.09 |
|               AwaitTaskAsync |  3.012 ns | 0.0103 ns | 0.0096 ns |   1.14 |
|     CompletedTaskResultAsync |  2.633 ns | 0.0070 ns | 0.0062 ns |   1.00 |
| CompletedTaskGetAwaiterAsync |  2.546 ns | 0.0045 ns | 0.0042 ns |   0.97 |

### BenchmarkDotNet

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Threading.Tasks;

namespace VSThreading301
{
    [BenchmarkDotNet.Attributes.DisassemblyDiagnoser(printAsm: false, printIL: true)]
    public class BenchCompletedTask
    {
        private volatile Task<int> _completedTask = Task.FromResult(1);
        const int Operations = 100000;

        [Benchmark(OperationsPerInvoke = Operations)]
        public async Task<int> AwaitTaskConfigureAwaitAsync()
        {
            await _completedTask.ConfigureAwait(false);
            int x = 0;
            for (int i = 0; i < Operations; i++)
            {
                x += await _completedTask.ConfigureAwait(false);
            }
            return x;
        }

        [Benchmark(OperationsPerInvoke = Operations)]
        public async Task<int> AwaitTaskAsync()
        {
            await _completedTask.ConfigureAwait(false);
            int x = 0;
            for (int i = 0; i < Operations; i++)
            {
                x += await _completedTask;
            }
            return x;
        }

        [Benchmark(OperationsPerInvoke = Operations, Baseline = true)]
        public async Task<int> CompletedTaskResultAsync()
        {
            await _completedTask.ConfigureAwait(false);
            int x = 0;
            for (int i = 0; i < Operations; i++)
            {
                if (!_completedTask.IsCompleted)
                    await _completedTask.ConfigureAwait(false);
                x += _completedTask.Result;
            }
            return x;
        }

        [Benchmark(OperationsPerInvoke = Operations)]
        public async Task<int> CompletedTaskGetAwaiterAsync()
        {
            await _completedTask.ConfigureAwait(false);
            int x = 0;
            for (int i = 0; i < Operations; i++)
            {
                if (!_completedTask.IsCompleted)
                    await _completedTask.ConfigureAwait(false);
                x += _completedTask.GetAwaiter().GetResult();
            }
            return x;
        }
    }

    public class Program
    {
        public static void Main(string[] args)
        {
            var summary = BenchmarkRunner.Run<BenchCompletedTask>();
        }
    }
}
```

**Comment by AArnott** [comment]
Have you filed a bug over at [dotnet/corefx](https://github.com/dotnet/corefx/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+configureawait+slow) for the slowness from `.ConfigureAwait(false)`? 400% slowdown seems incredibly high for a completed task.

Also, if you own the async method you're calling, given it usually is completed before returning have you considered using `ValueTask<T>` instead of `Task<T>`? That will avoid an allocation (although perhaps you're already caching a Task instance yourself) which can help throughput.

**Comment by KristianWedberg** [comment]
Good suggestion, I'll file it. 

Yes, I use cached tasks. And `ValueTask<T>` has certainly been a life save in some other places!

Cheers,
Kristian

**Comment by KristianWedberg** [comment]
Filed as https://github.com/dotnet/corefx/issues/30655, with additional tests and detail.

**Comment by sharwell** [comment]
:thought_balloon: A `CompletedResult` method which throws if the task is not completed would help write clear (readable) code and also avoid this issue.

**Comment by AArnott** [comment]
I like it. In fact I already have such an extension method (I think I called it `GetResultAssumeCompleted()`) for my own code clarify and enforcement. Adding such a method to vs-threading may be a good idea. But we might make the analyzer allow any method with that name regardless of where it comes from so that it can work on downlevel versions of VS.

**Comment by markusschaber** [comment]
We have another use case, apart from performance optimization: I use `Task.IsCompletedSuccessfully` to conditionally use `Task.Result` when I could already already use the result in case the task is finished, but I do not want to actively wait for the task to finish, as there are other things keeping the current task busy.
